src/Compilers/CSharp/Portable/Declarations/MergedTypeDeclaration.cs

```diff
@@ -123,6 +123,22 @@ public bool AnyMemberHasAttributes
             }
         }
 
+        public bool HasConstraints
+        {
+            get
+            {
+                foreach (var decl in this.Declarations)
```

Comment:
That's ok.
I'm not expert on allocations question, but my understanding is that the extensions we provide on `ImmutableArray` are good (they use a struct iterator).